File: ios/Parsers/GeoJSONParser.swift
import Foundation

struct GeoJSONParser {

  static func parse(data: Data, sourceType: String) throws -> [String: Any] {
    guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      throw GeoParserError.parseError("Invalid JSON")
    }

    let features = extractFeatures(from: json)

    var result: [String: Any] = ["type": "FeatureCollection", "sourceType": sourceType, "features": features]
    if let name = json["name"] as? String { result["name"] = name }
    if let props = json["properties"] as? [String: Any] {
      if let name = props["name"] as? String        { result["name"]        = name }
      if let desc = props["description"] as? String { result["description"] = desc }
    }
    return result
  }

  private static func extractFeatures(from json: [String: Any]) -> [[String: Any]] {
    guard let type = json["type"] as? String else { return [] }

    switch type {
    case "FeatureCollection":
      return (json["features"] as? [[String: Any]] ?? []).compactMap { parseFeature($0) }
    case "Feature":
      return parseFeature(json).map { [$0] } ?? []
    default:
      if isGeometryType(type), let feature = bareGeometryToFeature(json) { return [feature] }
      return []
    }
  }

  private static func parseFeature(_ json: [String: Any]) -> [String: Any]? {
    guard json["type"] as? String == "Feature",
          let geometry = json["geometry"] as? [String: Any]
    else { return nil }

    let properties = json["properties"] as? [String: Any] ?? [:]
    var feature: [String: Any] = ["type": "Feature", "geometry": geometry, "properties": properties]
    if let id = json["id"] { feature["id"] = "\(id)" }
    return feature
  }

  private static func bareGeometryToFeature(_ json: [String: Any]) -> [String: Any]? {
    guard json["type"] as? String != nil else { return nil }
    return ["type": "Feature", "geometry": json, "properties": [:] as [String: Any]]
  }

  private static func isGeometryType(_ type: String) -> Bool {
    let types: Set<String> = [
      "Point", "LineString", "Polygon",
      "MultiPoint", "MultiLineString", "MultiPolygon",
      "GeometryCollection"
    ]
    return types.contains(type)
  }
}

File: android/src/main/java/expo/modules/geoparser/parsers/GeoJSONParser.kt
package expo.modules.geoparser.parsers

import expo.modules.geoparser.GeoParserError
import org.json.JSONArray
import org.json.JSONObject
import java.io.InputStream

object GeoJSONParser {

    fun parse(input: InputStream, sourceType: String, onFeatures: (List<Map<String, Any?>>, isLast: Boolean) -> Unit): Map<String, Any?> {
        val json = try {
            JSONObject(input.reader(Charsets.UTF_8).readText())
        } catch (e: Exception) {
            throw GeoParserError.ParseError("Invalid JSON")
        }

        val features = extractFeatures(json)
        val batchSize = 200
        if (features.isEmpty()) {
            onFeatures(emptyList(), true)
        } else {
            var idx = 0
            while (idx < features.size) {
                val end = minOf(idx + batchSize, features.size)
                onFeatures(features.subList(idx, end), end >= features.size)
                idx = end
            }
        }

        val result = mutableMapOf<String, Any?>(
            "type" to "FeatureCollection",
            "sourceType" to sourceType
        )
        json.optString("name").takeIf { it.isNotEmpty() }?.let { result["name"] = it }
        json.optJSONObject("properties")?.let { props ->
            props.optString("name").takeIf { it.isNotEmpty() }?.let { result["name"] = it }
            props.optString("description").takeIf { it.isNotEmpty() }?.let { result["description"] = it }
        }

        return result
    }

    private fun extractFeatures(json: JSONObject): List<Map<String, Any?>> {
        return when (val type = json.optString("type")) {
            "FeatureCollection" -> {
                val arr = json.optJSONArray("features") ?: return emptyList()
                (0 until arr.length()).mapNotNull { parseFeature(arr.getJSONObject(it)) }
            }
            "Feature" -> listOfNotNull(parseFeature(json))
            else -> {
                if (isGeometryType(type)) listOfNotNull(bareGeometryToFeature(json))
                else emptyList()
            }
        }
    }

    private fun parseFeature(json: JSONObject): Map<String, Any?>? {
        if (json.optString("type") != "Feature") return null
        val geometry = json.optJSONObject("geometry") ?: return null
        val properties = json.optJSONObject("properties") ?: JSONObject()
        val feature = mutableMapOf<String, Any?>(
            "type" to "Feature",
            "geometry" to toMap(geometry),
            "properties" to toMap(properties)
        )
        json.opt("id")?.takeIf { it != JSONObject.NULL }?.let { feature["id"] = it.toString() }
        return feature
    }

    private fun bareGeometryToFeature(json: JSONObject): Map<String, Any?>? {
        if (json.optString("type").isEmpty()) return null
        return mapOf(
            "type" to "Feature",
            "geometry" to toMap(json),
            "properties" to emptyMap<String, Any?>()
        )
    }

    private fun isGeometryType(type: String): Boolean = type in setOf(
        "Point", "LineString", "Polygon",
        "MultiPoint", "MultiLineString", "MultiPolygon",
        "GeometryCollection"
    )

    fun toMap(obj: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        for (key in obj.keys()) map[key] = toKotlin(obj.get(key))
        return map
    }

    private fun toList(arr: JSONArray): List<Any?> =
        (0 until arr.length()).map { toKotlin(arr.get(it)) }

    private fun toKotlin(value: Any?): Any? = when (value) {
        is JSONObject -> toMap(value)
        is JSONArray -> toList(value)
        JSONObject.NULL -> null
        else -> value
    }
}
